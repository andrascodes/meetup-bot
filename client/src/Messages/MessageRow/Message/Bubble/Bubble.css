.Bubble {
  margin: auto;
  margin-bottom: 0.25em;
  border: 0.1em solid;
  border-radius: 1em;
  border-color: rgba(0, 0, 0, 0.1);
  padding: 0.7em 1em 0.7em 1em;
}

.ErrorBubble {
  border: 0.25em solid rgb(255, 102, 102);
}

.Incoming {
  background-color: #f1f0f0;
}

.Outgoing {
  background-color: #0084ff;
  color: #fff;
}

.ImageMessage {
  border-radius: 1em;
}