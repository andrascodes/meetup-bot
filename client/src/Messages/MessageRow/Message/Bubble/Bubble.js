import React from 'react'
import ReactToolTip from 'react-tooltip'

import { ListMessage } from './ListMessage'
import { GenericMessage } from './GenericMessage'
import { ButtonMessage } from './ButtonMessage'

import './Bubble.css'

export const Bubble = props => {
  const direction = (props.direction === 'incoming') ? 'Incoming' : 'Outgoing'
  const tooltipTag = (props.error) ? { 'data-tip': 'React tooltip' } : {}

  let error = ''
  let Tooltip = ''
  if(props.error) {
    error = 'ErrorBubble'
    Tooltip = (
      <div className="ErrorTooltip">
        <ReactToolTip place="top" type="dark" effect="solid">
          {props.response.error.message}
        </ReactToolTip>
      </div>
    )
  }

  const mapMessageTypeToComponent = (type, message, text) => {
    if(type === 'text') {
      return (<div className="TextMessage">{(message && message.message && message.message.text) || text}</div>)
    }
    else if(type === 'quick_reply') {
      return (<div className="QuickReplyMessage">
                {`${message.message.text} (Payload: ${message.message.quick_reply.payload})`}
              </div>)
    }
    else if(type === 'postback' || type === 'referral_existing' || type === 'referral_new') {
      return (<div className="PostbackMessage">
                {`POSTBACK (Payload: ${message.postback.payload})`}
              </div>)
    }
    else if(type === 'list') {
      return (<ListMessage {...message.message} />)
    }
    else if(type === 'generic') {
      return (<GenericMessage {...message.message} />)
    }
    else if(type === 'button') {
      return(
        <ButtonMessage {...message.message}/>
      )
    }
    else if(type === 'image' || type === 'sticker') {
      return (
        message.message.attachments.map((img, ind) => (
          <img src={img.payload.url} alt="" key={ind} className="ImageMessage img-responsive"/>
        ))
      )
    }
    else if(type === 'audio' || type === 'file' || type === 'video') {
      const typeToPropertiesMap = {
        audio: {
          className: 'AudioMessage',
          text: 'Audio message',
          icon: 'fa-volume-up'
        },
        file: {
          className: 'FileMessage',
          text: 'File message',
          icon: 'fa-file-o'
        },
        video: {
          className: 'VideoMessage',
          text: 'Video message',
          icon: 'fa-video-camera'
        }
      }
      const properties = typeToPropertiesMap[type]

      return (
        message.message.attachments.map((file, ind) => (
          <div key={ind} className={`MediaMessage ${properties.className}`}>
            <i className={`fa ${properties.icon}`} />
            {` ${properties.text} `}
            <a href={file.payload.url} target="_blank">
              <i className="fa fa-download" />
            </a>
          </div>
        ))
      )
    }
    else if(type === 'location') {
      return (
          message.message.attachments.map((loc, ind) => (
            <div className="LocationMessage" key={ind}>
              <a 
                className="LocationTitle" 
                href={`https://www.google.com/maps/?q=${loc.payload.coordinates.lat},${loc.payload.coordinates.long}`}
                target="_blank"
              >
                {`${loc.title}:`}
              </a>
              <div className="LocationCoordinates">
                {`Latitude: ${loc.payload.coordinates.lat}`}
                <br/>
                {`Longitude: ${loc.payload.coordinates.long}`}
              </div>
            </div>
          ))
      )
    }
    else {
      return(
        <div className="UnknownMessage">
          {props.text}
        </div>
      )
    }
  }

  const message = mapMessageTypeToComponent(props.type, props.message, props.text)
  
  return (
    <div>
      <div className={`${direction} Bubble ${error}`} {...tooltipTag} >
        {message}
      </div>
      {Tooltip}
    </div>
  )
}