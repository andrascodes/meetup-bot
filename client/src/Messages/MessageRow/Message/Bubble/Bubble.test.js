import React from 'react'
import renderer from 'react-test-renderer'

import { Bubble } from './Bubble'

describe.only('Bubble', () => {

  test('renders incoming Text Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'text'}
        direction={'incoming'} 
        message={{
          message: {
            text: 'Incoming text message'
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  test('renders outgoing Text Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'text'}
        direction={'outgoing'} 
        message={{
          message: {
            text: 'Outgoing text message'
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  test('renders Error tooltip correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'text'}
        direction={'outgoing'} 
        message={{
          message: {
            text: 'Outgoing text message'
          }
        }}
        error={true}
        response={{
          error: {
            message: 'Something went wrong'
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  test('renders Quick Reply Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'quick_reply'}
        direction={'outgoing'} 
        message={{
          message: {
            text: 'Quick Reply',
            quick_reply: {
              payload: 'PAYLOAD'
            }
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  test('renders Attachment Bubbles correctly', () => {
    let tree = renderer.create(
      <Bubble 
        type={'image'}
        direction={'incoming'} 
        message={{
          message: {
            attachments: [
              {
                payload: {
                  url: 'https://www.google.com'
                }
              }
            ]
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()

    tree = renderer.create(
      <Bubble 
        type={'audio'}
        direction={'incoming'} 
        message={{
          message: {
            attachments: [
              {
                payload: {
                  url: 'https://www.google.com'
                }
              }
            ]
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
    
    tree = renderer.create(
      <Bubble 
        type={'file'}
        direction={'incoming'} 
        message={{
          message: {
            attachments: [
              {
                payload: {
                  url: 'https://www.google.com'
                }
              }
            ]
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
    
    tree = renderer.create(
      <Bubble 
        type={'video'}
        direction={'incoming'} 
        message={{
          message: {
            attachments: [
              {
                payload: {
                  url: 'https://www.google.com'
                }
              }
            ]
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  test('renders Location Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'location'}
        direction={'incoming'} 
        message={{
          message: {
            attachments: [
              {
                payload: {
                  coordinates: {
                    lat: '42.34',
                    long: '-35.45',
                  },
                  title: 'Location'
                }
              }
            ]
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders Unknown Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'unknown'}
        text={'Unknown message placeholder'}
        direction={'outgoing'}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders List Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'list'}
        direction={'outgoing'} 
        message={{
          message: {
            attachment: {
              payload: {
                elements: [
                  {
                    title: 'Title 1',
                    subtitle: 'Subtitle 1',
                    default_action: {
                      url: 'https://www.google.com'
                    },
                    image_url: 'https://www.google.com',
                    buttons: [
                      {
                        title: 'ElemBtn 1'
                      }
                    ]
                  },
                  {
                    title: 'Title 2',
                    subtitle: 'Subtitle 2',
                    default_action: {
                      url: 'https://www.google.com'
                    },
                    image_url: 'https://www.google.com',
                    buttons: [
                      {
                        title: 'ElemBtn 2'
                      }
                    ]
                  },
                ],
                buttons: [
                  {
                    title: 'ListBtn'
                  }
                ]
              }
            }
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  test('renders Generic Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'generic'}
        direction={'outgoing'} 
        message={{
          message: {
            attachment: {
              payload: {
                elements: [
                  {
                    image_url: 'https://www.google.com',
                    title: 'Title1',
                    subtitle: 'Subtitle1',
                    default_action: {
                      url: 'https://www.google.com'
                    },
                    buttons: [
                      {
                        title: 'Btn1 title1',
                      },
                      {
                        title: 'Btn1 title2',
                      },
                    ]
                  },
                  {
                    image_url: 'https://www.google.com',
                    title: 'Title2',
                    subtitle: 'Subtitle2',
                    default_action: {
                      url: 'https://www.google.com'
                    },
                    buttons: [
                      {
                        title: 'Btn2 title1',
                      },
                      {
                        title: 'Btn2 title2',
                      },
                    ]
                  }
                ]
              }
            }
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('renders Button Bubble correctly', () => {
    const tree = renderer.create(
      <Bubble 
        type={'button'}
        direction={'outgoing'} 
        message={{
          message: {
            attachment: {
              payload: {
                text: 'Message Text',
                buttons: [
                  {
                    title: 'Btn title1',
                  },
                  {
                    title: 'Btn title2',
                  },
                ]
              }
            }
          }
        }}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

})