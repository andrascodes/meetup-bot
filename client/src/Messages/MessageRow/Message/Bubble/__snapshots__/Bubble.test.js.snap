exports[`Bubble renders Attachment Bubbles correctly 1`] = `
<div>
  <div
    className="Incoming Bubble ">
    <img
      alt=""
      className="ImageMessage img-responsive"
      src="https://www.google.com" />
  </div>
  
</div>
`;

exports[`Bubble renders Attachment Bubbles correctly 2`] = `
<div>
  <div
    className="Incoming Bubble ">
    <div
      className="MediaMessage AudioMessage">
      <i
        className="fa fa-volume-up" />
       Audio message 
      <a
        href="https://www.google.com"
        target="_blank">
        <i
          className="fa fa-download" />
      </a>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Attachment Bubbles correctly 3`] = `
<div>
  <div
    className="Incoming Bubble ">
    <div
      className="MediaMessage FileMessage">
      <i
        className="fa fa-file-o" />
       File message 
      <a
        href="https://www.google.com"
        target="_blank">
        <i
          className="fa fa-download" />
      </a>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Attachment Bubbles correctly 4`] = `
<div>
  <div
    className="Incoming Bubble ">
    <div
      className="MediaMessage VideoMessage">
      <i
        className="fa fa-video-camera" />
       Video message 
      <a
        href="https://www.google.com"
        target="_blank">
        <i
          className="fa fa-download" />
      </a>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Button Bubble correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ">
    <div
      className="ButtonMessage">
      <div
        className="ButtonMessageText ButtonDataContainer">
        Message Text
      </div>
      <div
        className="ButtonMessageButton ButtonDataContainer text-center">
        Btn title1
      </div>
      <div
        className="ButtonMessageButton ButtonDataContainer text-center">
        Btn title2
      </div>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Error tooltip correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ErrorBubble"
    data-tip="React tooltip">
    <div
      className="TextMessage">
      Outgoing text message
    </div>
  </div>
  <div
    className="ErrorTooltip">
    <div
      className="__react_component_tooltip place-top type-dark "
      data-id="tooltip">
      
    </div>
  </div>
</div>
`;

exports[`Bubble renders Generic Bubble correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ">
    <div
      className="GenericMessage row">
      <div
        className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-left">
        <button
          className="GenericPagingButton btn disabled"
          onClick={[Function]}>
          <i
            className="fa fa-chevron-left" />
        </button>
      </div>
      <div
        className="col-lg-8 col-md-8 col-sm-8 col-xs-8 text-center">
        <div
          className="GenericElement">
          <img
            alt=""
            className="GenericElementImage img-responsive"
            src="https://www.google.com" />
          <div
            className="GenericElementTitleContainer GenericDataContainer text-left">
            <div
              className="GenericElementTitle">
              Title1
            </div>
            <div
              className="GenericElementSubtitle">
              Subtitle1
            </div>
            <div
              className="GenericElementDefaultAction GenericElementSubtitle">
              www.google.com
            </div>
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn1 title1
          </div>
          <div
            className="GenericElementButton GenericDataContainer text-center">
            Btn1 title2
          </div>
        </div>
      </div>
      <div
        className="col-lg-2 col-md-2 col-sm-2 col-xs-2 text-right">
        <button
          className="GenericPagingButton btn "
          onClick={[Function]}>
          <i
            className="fa fa-chevron-right" />
        </button>
      </div>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders List Bubble correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ">
    <div
      className="ListMessage">
      <div
        className="ListElementContainer">
        <div
          className="ListElement row">
          <div
            className="col-lg-6 col-md-6 col-sm-6 col-xs-6 text-left">
            <div
              className="ListElementTitle">
              Title 1
            </div>
            <div
              className="ListElementSubtitle">
              Subtitle 1
            </div>
            <div
              className="ListElementDefaultAction ListElementSubtitle">
              www.google.com
            </div>
            <div
              className="ListElementButton">
              ElemBtn 1
            </div>
          </div>
          <div
            className="col-lg-6 col-md-6 col-sm-6 col-xs-6 text-right">
            <img
              alt=""
              className="ListElementImage"
              src="https://www.google.com" />
          </div>
        </div>
      </div>
      <div
        className="ListElementContainer">
        <div
          className="ListElement row">
          <div
            className="col-lg-6 col-md-6 col-sm-6 col-xs-6 text-left">
            <div
              className="ListElementTitle">
              Title 2
            </div>
            <div
              className="ListElementSubtitle">
              Subtitle 2
            </div>
            <div
              className="ListElementDefaultAction ListElementSubtitle">
              www.google.com
            </div>
            <div
              className="ListElementButton">
              ElemBtn 2
            </div>
          </div>
          <div
            className="col-lg-6 col-md-6 col-sm-6 col-xs-6 text-right">
            <img
              alt=""
              className="ListElementImage"
              src="https://www.google.com" />
          </div>
        </div>
      </div>
      <div
        className="ListButton text-center">
        ListBtn
      </div>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Location Bubble correctly 1`] = `
<div>
  <div
    className="Incoming Bubble ">
    <div
      className="LocationMessage">
      <a
        className="LocationTitle"
        href="https://www.google.com/maps/?q=42.34,-35.45"
        target="_blank">
        undefined:
      </a>
      <div
        className="LocationCoordinates">
        Latitude: 42.34
        <br />
        Longitude: -35.45
      </div>
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Quick Reply Bubble correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ">
    <div
      className="QuickReplyMessage">
      Quick Reply (Payload: PAYLOAD)
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders Unknown Bubble correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ">
    <div
      className="UnknownMessage">
      Unknown message placeholder
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders incoming Text Bubble correctly 1`] = `
<div>
  <div
    className="Incoming Bubble ">
    <div
      className="TextMessage">
      Incoming text message
    </div>
  </div>
  
</div>
`;

exports[`Bubble renders outgoing Text Bubble correctly 1`] = `
<div>
  <div
    className="Outgoing Bubble ">
    <div
      className="TextMessage">
      Outgoing text message
    </div>
  </div>
  
</div>
`;
